File: labs/lab3/9771_Sachu/sync.py
from flask import Flask, request, jsonify
from flask_cors import CORS
from pymongo import MongoClient
import json

app = Flask(__name__)
CORS(app)  # This will enable CORS for all routes

# MongoDB connection setup
username = 'i40'  # Your MongoDB username
password = 'dbms2'  # Your MongoDB password
host = 'cluster0.lixbqmp.mongodb.net'  # Your MongoDB host
database = 'lab3'  # Your database name

# Creating MongoDB client
client = MongoClient(f'mongodb+srv://{username}:{password}@{host}/{database}?tls=true&tlsInsecure=true')
db = client[database]
collection = db['9771_Sachu']  # Your collection name

@app.route('/sync/to/mongodb', methods=['POST'])
def sync_to_mongodb():
    data = request.get_json()

    if not data or 'documents' not in data or not isinstance(data['documents'], list) or len(data['documents']) == 0:
        return jsonify({"error": "Documents must be a non-empty list."}), 400

    try:
        collection.insert_many(data['documents'])
        return jsonify({"message": "Data synced to MongoDB successfully!"}), 200
    except Exception as e:
        return jsonify({"error": str(e)}), 500

if __name__ == '__main__':
    app.run(debug=True)

File: labs/lab3/shankar4588/sync.js
const express = require('express');
const bodyParser = require('body-parser');
const cors = require('cors'); // Import CORS
const MongoClient = require('mongodb').MongoClient;

const app = express();
const port = 3005;

// MongoDB connection URI
const mongoURI = "mongodb+srv://i40:dbms2@cluster0.lixbqmp.mongodb.net/lab3?retryWrites=true&w=majority";
const mongoCollection = "shankar_data";  // Collection name

// Middleware
app.use(cors()); // Use CORS middleware
app.use(bodyParser.json());

// Route for root '/'
app.get('/', (req, res) => {
    res.send('Welcome to the Sync Server!');
});

// Sync route to handle POST requests from IndexedDB
app.post('/sync', async (req, res) => {
    try {
        const client = new MongoClient(mongoURI, { useNewUrlParser: true, useUnifiedTopology: true });
        await client.connect();
        const db = client.db('lab3');  // Database name
        const collection = db.collection(mongoCollection);  // Collection name

        const indexedDBData = req.body; // Data sent from IndexedDB

        console.log('Received data:', indexedDBData);  // Log received data

        const result = await collection.insertMany(indexedDBData);

        console.log(`${result.insertedCount} documents inserted.`);
        res.status(200).send(`${result.insertedCount} documents successfully synced to MongoDB.`);
        
        await client.close();  // Close the connection
    } catch (error) {
        console.error('Sync error:', error);
        res.status(500).send('Error syncing data');
    }
});

// Start the server
app.listen(port, () => {
    console.log(`Server running on port ${port}`);
});
